File: urbanairship-automation/src/main/java/com/urbanairship/automation/NetworkMonitor.java
package com.urbanairship.automation;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Build;

import com.urbanairship.Logger;
import com.urbanairship.UAirship;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
public class NetworkMonitor {

    public interface ConnectionListener {

        /**
         * Called when the connection state changed.
         * @param isConnected
         */
        void onConnectionChanged(boolean isConnected);
    }

    private ConnectionListener connectionListener;

    private final ConnectivityManager.NetworkCallback networkCallback = new ConnectivityManager.NetworkCallback() {
        @Override
        public void onAvailable(@NonNull Network network) {
            if (connectionListener != null) {
                connectionListener.onConnectionChanged(true);
            }
        }

        public void onLost(@NonNull Network network) {
            if (connectionListener != null) {
                connectionListener.onConnectionChanged(false);
            }
        }
    };

    public void registerNetworkCallback() {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) UAirship.getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                connectivityManager.registerDefaultNetworkCallback(networkCallback);
            } else {
                NetworkRequest request = new NetworkRequest.Builder()
                        .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET).build();
                connectivityManager.registerNetworkCallback(request, networkCallback);
            }
        } catch (SecurityException e) {
            // Workaround "Package android does not belong to 10246" exceptions that occur on
            // a small set of devices.
            Logger.warn(e, "NetworkMonitor failed to register network callback!");
        }
    }

    public void unregisterNetworkCallback() {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) UAirship.getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);

        try {
            connectivityManager.unregisterNetworkCallback(networkCallback);
        } catch (Exception e) {
            // Ignore unregister errors in case the callback wasn't previously registered.
            Logger.warn(e, "NetworkMonitor failed to unregister network callback!");
        }
    }

    public void setConnectionListener(ConnectionListener listener) {
        connectionListener = listener;
        registerNetworkCallback();
    }

    public void teardown() {
        connectionListener = null;
        unregisterNetworkCallback();
    }
}

File: urbanairship-automation/src/main/java/com/urbanairship/automation/rewrite/utils/NetworkMonitor.kt
package com.urbanairship.automation.rewrite.utils

import android.content.Context
import android.content.Context.CONNECTIVITY_SERVICE
import android.net.ConnectivityManager
import android.net.ConnectivityManager.NetworkCallback
import android.net.NetworkCapabilities.NET_CAPABILITY_INTERNET
import android.net.NetworkRequest
import android.os.Build
import com.urbanairship.UALog
import com.urbanairship.automation.rewrite.DerivedStateFlow
import com.urbanairship.util.Network
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.callbackFlow

internal class NetworkMonitor(
    context: Context,
    network: Network = Network.shared()
) {
    var isConnected: StateFlow<Boolean> = DerivedStateFlow(
        onValue = { network.isConnected(context) },
        updates = callbackFlow {
            val callback = object : NetworkCallback() {
                override fun onAvailable(network: android.net.Network) {
                    this@callbackFlow.trySend(true)
                }

                override fun onLost(network: android.net.Network) {
                    this@callbackFlow.trySend(false)
                }
            }

            val service = context.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    service.registerDefaultNetworkCallback(callback)
                } else {
                    val networkRequest = NetworkRequest.Builder()
                        .addCapability(NET_CAPABILITY_INTERNET)
                        .build()
                    service.registerNetworkCallback(networkRequest, callback)
                }
            } catch (ex: Exception) {
                UALog.e(ex) { "Failed to subscribe for network status update" }
                this@callbackFlow.trySend(true)
            }

            trySend(network.isConnected(context))
            awaitClose { service.unregisterNetworkCallback(callback) }
        }
    )

    companion object {
        fun shared(context: Context): NetworkMonitor {
            return NetworkMonitor(context)
        }
    }
}
